File: platforms/ios/runtime/Sources/NativeblocksRuntime/localization/LocalizationUseCase.swift
import Foundation
import NativeblocksRuntimeFFI

internal final class LocalizationUseCase {

    private let runtimeClient: NativeRuntimeClientManager
    private let stateManager: LocalizationStateManager

    init(runtimeClient: NativeRuntimeClientManager, stateManager: LocalizationStateManager) {
        self.runtimeClient = runtimeClient
        self.stateManager = stateManager
    }

    private var client: LocalizationClient {
        return runtimeClient.localizationClient
    }

    func setLocalization(languageCode: String) {
        client.setLanguageCode(languageCode: languageCode)
        Task.detached { [client] in
            try? await client.syncLocalization(languageCode: languageCode)
        }
    }

    func translate(key: String) -> String? {
        return stateManager.translate(key: key)
    }
}

File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/localization/LocalizationUseCase.kt
package io.nativeblocks.runtime.localization

import io.nativeblocks.runtime.engine.LocalizationStateManager
import io.nativeblocks.runtime.engine.NativeEngineClientManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

internal class LocalizationUseCase(
    private val engineClient: NativeEngineClientManager,
    private val stateManager: LocalizationStateManager,
) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val client get() = engineClient.localizationClient

    fun setLocalization(languageCode: String) {
        scope.launch {
            runCatching { client.setLanguageCode(languageCode) }
            runCatching { client.syncLocalization(languageCode) }
        }
    }

    fun translate(key: String): Result<String?> {
        return runCatching { stateManager.translate(key) }
    }
}
